### PR TITLE
[SPARK-41128][CONNECT][PYTHON] Implement `DataFrame.fillna ` and `DataFrame.na.fill `

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement `DataFrame.fillna ` and `DataFrame.na.fill`

### Why are the changes needed?
For API coverage

### Does this PR introduce _any_ user-facing change?
yes, new methods

### How was this patch tested?
added UT

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @cloud-fan @amaliujia @grundprinzip

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -316,6 +319,36 @@ message StatCrosstab {
...
+  Relation input = 1;
+
+  // (Optional) Optional list of column names to consider.
+  repeated string cols = 2;
+
+  // (Required) Values to replace null values with. Should contains at least 1 item.
+  repeated ValueType values = 3;
+
+  // Available data types.
+  message ValueType {
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: Is this `value` or `type` which probably different things? If it is just value then just call it value (same for type).

File: python/pyspark/sql/tests/connect/test_connect_plan_only.py [hunk 2]
```diff
@@ -70,6 +70,37 @@ def test_filter(self):
...
+        )
+        self.assertEqual(len(plan.root.fill_na.values), 1)
+        self.assertEqual(plan.root.fill_na.values[0].bool_value, True)
+        self.assertEqual(plan.root.fill_na.cols, ["col_a", "col_b", "col_c"])
+
+        plan = df.fillna({"col_a": 1.5, "col_b": "abc"})._plan.to_proto(self.connect)
+        self.assertEqual(len(plan.root.fill_na.values), 2)
+        self.assertEqual(plan.root.fill_na.values[0].double_value, 1.5)
+        self.assertEqual(plan.root.fill_na.values[1].string_value, "abc")
+        self.assertEqual(plan.root.fill_na.cols, ["col_a", "col_b"])
```
Comment (amaliujia) [inline python/pyspark/sql/tests/connect/test_connect_plan_only.py:102]: Do we need at least one e2e test or such test suite is enough?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: yes, should be `type`s here

File: python/pyspark/sql/tests/connect/test_connect_plan_only.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/tests/connect/test_connect_plan_only.py:102]: we don't have e2e test for most methods, but let me add one for this method

Comment (amaliujia) [reply python/pyspark/sql/tests/connect/test_connect_plan_only.py:102]: Yeah I am actually we maybe needs at least one for each API to make sure they pass through.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -316,6 +319,36 @@ message StatCrosstab {
...
+  Relation input = 1;
+
+  // (Optional) Optional list of column names to consider.
+  repeated string cols = 2;
+
+  // (Required) Values to replace null values with. Should contains at least 1 item.
+  repeated Type values = 3;
+
+  // Available data types.
+  message Type {
```
Comment (HyukjinKwon) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: Could we probably use `Literal` to ship these values? We can call `eval` to take this out in JVM side.

Comment (HyukjinKwon) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: Or is this consistent with other API we added so far? Then I am good.

Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: I think this is the first method that only support several datatype.

> Could we probably use Literal to ship these values?

the question is do we need to limit the datatypes in proto message?

Comment (HyukjinKwon) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: Okay, I am fine with the current way for now.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by amaliujia** [comment]
LGTM

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 3 above)
Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: @HyukjinKwon thank you. There are also some other methods only supporting a subset of datetypes. If we prefer use `Literal` then I can make such change in the future.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: It seems more flexible to restrict the types in the server side. Once we relax the restriction in the future, people don't need to update the clients.

Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:342]: @cloud-fan @HyukjinKwon make sense, let me update it

Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto]: updated

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 4]
```diff
@@ -226,6 +226,58 @@ package object dsl {
       }
     }
 
+    implicit class DslNAFunctions(val logicalPlan: Relation) {
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:229]: no, not needed, but `implicit` is already widely used in both catalyst and connect...

**Comment by zhengruifeng** [comment]
merged into master, thank you guys

**Pull request closed without merging**